py/tests/integration/test_spend_tx.py

```diff
@@ -88,6 +79,71 @@ def test_successful():
     common.stop_node(alice_node)
     shutil.rmtree(root_dir)
 
+def test_not_enough_tokens():
+    # Bob should not be able to send more tokens than he has
+    #
+    # Let's say Bob has 100 tokens. He should not be able to send more than
+    # 100 tokens to Alice.
+    #
+    # If there's an incoming but unconfirmed deposit into Bob's account then Bob
+    # should not be able to use the incoming tokens until the spend transaction
+    # they are in is confirmed.
+    test_settings = settings["test_not_enough_tokens"]
+    coinbase_reward = common.coinbase_reward() 
+    (root_dir, bob_node, bob_api, bob_top) = setup_node_with_tokens(test_settings, "bob") 
+
+    spend_tx_amt = test_settings["spend_tx"]["amount"]
+    spend_tx_fee = test_settings["spend_tx"]["fee"]
+    bob_balance = common.get_account_balance(bob_api)
+    bob_has_not_enough_tokens = bob_balance.balance < spend_tx_amt + spend_tx_fee
+    assert_equals(bob_has_not_enough_tokens, True)
+    print("Bob initial balance is " + str(bob_balance.balance) +
+            " and he will try to spend " + str(spend_tx_amt + spend_tx_fee))
+
+    alice_balance0 = common.get_account_balance(bob_api, pub_key=test_settings["spend_tx"]["alice_pubkey"])
+
+    # Bob tries to send some tokens to Alice
+    bob_internal_api = common.internal_api(bob_node)
+    spend_tx_obj = SpendTx(
+        recipient_pubkey=test_settings["spend_tx"]["alice_pubkey"],
+        amount=spend_tx_amt,
+        fee=spend_tx_fee)
+    print("Bob's spend_tx is " + str(spend_tx_obj))
+    bob_internal_api.post_spend_tx(spend_tx_obj)
+    print("Transaction sent")
+
+    # ensure Bob balance had not changed
+    bob_balance1 = common.get_account_balance(bob_api)
+    assert_equals(bob_balance.balance, bob_balance1.balance)
+
+    # ensure Alice balance had not changed
+    alice_balance1 = common.get_account_balance(bob_api, pub_key=test_settings["spend_tx"]["alice_pubkey"])
+    # wait for a block to be mined
+    print("Waiting for a next block to be mined")
+    common.wait_until_height(bob_api, bob_top.height + 1)
+
+    bob_new_top = bob_api.get_top()
+    bob_new_balance = common.get_account_balance(bob_api)
+    
+    blocks_mined = bob_new_top.height - bob_top.height
+
+    # Since Bob had mined the block he is receiving the coinbase_reward
+    expected_balance = bob_balance.balance + coinbase_reward * blocks_mined
+    assert_equals(bob_new_balance.balance, expected_balance)
+
+    alice_balance = Balance(balance=0) 
```

Comment:
This is not needed as variable re-assigned in next line (136).  And symbol `Balance` is not imported hence causing failure.
I am preparing patch locally.

Reply:
@velzevur  https://github.com/velzevur/epoch/pull/1